simapp/app.go

```diff
@@ -498,6 +497,7 @@ func NewSimApp(
 
 	// initialize BaseApp
 	app.SetInitChainer(app.InitChainer)
+	app.SetPreBlocker(app.PreBlocker)
```

Comment:
Sorry for my bad desc, the hardcode `RunMigrationBeginBlock` was introduced in https://github.com/cosmos/cosmos-sdk/pull/16583, we decide to handle with more generic `PreBlock` which is cleaner and could be reuse in the future.

Reply:
> So you're introducing a new life cycle event in this PR? Could you let us know why? It's a bit hard for me to understand the problem being solved from the 2 bullet points in the PR desc.

## Context

### The Original Problem

When upgrading to sdk 0.47, the storage format for consensus parameters changed, but in the migration block, `ctx.ConsensusParams()` is always `nil`, because it fails to load the old format using new code, it's supposed to be migrated by the `x/upgrade` module first, but unfortunately, the migration happens in `BeginBlocker` handler, which runs after the `ctx` is initialized.
When we try to solve this, we find the `x/upgrade` module can't modify the context to make the consensus parameters visible for the other modules, the context is passed by value, and sdk team want to keep it that way, that's good for isolations between modules.
The first alternative solution introduced a `MigrateModuleManager`, which only includes the `x/upgrade` module right now, and baseapp will run their `BeginBlocker`s before the other modules, and reload context's consensus parameters in between, but I think it's too hacky, and suggested this new lifecycle method.

### `PreBlocker`

There are two semantics around the new lifecycle method:

- It runs before the `BeginBlocker` of all modules
- It can modify consensus parameters in storage, and signal the caller through the return value.

When it returns `ConsensusParamsChanged=true`, the caller must refresh the consensus parameter in the context:
```
ctx = ctx.WithConsensusParams(app.GetConsensusParams())
```
and pass the new `ctx` to `BeginBlocker` methods, it should also update the `deliverState.ctx`:
```
app.deliverState.ctx = ctx
```
So the changes are visible to all the other lifecycle methods.